docs/conceptual/runtime/weight.md

```diff
@@ -0,0 +1,120 @@
+  means that makes sense to that particular chain. An example of such dispatches
+  in substrate is the call to a smart contract (todo: link when this is
+  implemented).
+
+Given the above details, one can write their own weight computation function,
+`$x` in the above listing, to determine the weight of dispatchable. The
+programming details of doing this is beyond the scope of this document. For more
+information on that, refer to:
+- the [weight
+  module's](https://crates.parity.io/sr_primitives/weights/index.html)
+  documentation in substrate-primitives.
+- the
+  [`srml-example`](https://github.com/paritytech/substrate/blob/master/srml/example/src/lib.rs)
+  module and the inline documentation. Skim the code for `#[weight]` attributes
+  and all the related documentations.
+- the section on weights in Substrate recipes. TODO
+
+The system module is responsible for accumulating the weight of each block as it
+gets executed and making sure that it does not exceed the limit. The
+transaction-payment module is responsible for interpreting these weights and
+deducting fees based upon them.
+
+
+## More on Weights: Block Weight and Length Limit
+
+Aside from affecting the fees, the main purpose of the weight system is to
+prevent a block from being too filled with transactions. The system module,
+while processing blocks≠transactions within a block, accumulates both the total
+length of the block (sum of encoded transactions in number of bytes) and the
+total weight of the blocks. At any points, if these numbers surpass the limits,
+no further transactions are accepted to the block. These limits are defined in
+[`MaximumBlockLength`](https://crates.parity.io/srml_system/trait.Trait.html#associatedtype.MaximumBlockLength)
+and
+[`MaximumBlockWeight`](https://crates.parity.io/srml_system/trait.Trait.html#associatedtype.MaximumBlockLength)
+respectively.
+
+One important note about these limits is that a ratio of them are reserved for
+the `Operational` dispatch class. This rule applies to both of the limits and
+the ratio can be found in
+[`AvailableBlockRatio`](https://crates.parity.io/srml_system/trait.Trait.html#associatedtype.AvailableBlockRatio).
+For example, if the block length limit is 1 mega bytes and the ratio is set the
+80%, all transactions can fill the first 800 kilo bytes of the block while the
+last 200 can only be filled by the operation class.
+
+
+## Frequent Questions
+
+> _Some operations get more expensive as storage grows and shrinks. How can the
+> weight system represent that?_
+
+As mentioned, weight should be known _readily_ and _ahead of dispatch_. Peeking
+the storage and analyzing it just to determine how much weight might be consumed
+does not fit this definition quite well. In this case the implementation of the
+dispatch should take the state of the change into account and manually take
+extra fees, bonds or take any other measures to make sure that the transaction
+is safe.
+
```

Comment:
@kianenigma I had also thought of this solution. What do you think?

The transaction contains, as a parameter, the expected value from storage. Since the value is a parameter, the weight can be calculated in advance. When the transaction is dispatched, it `ensure!`s that the parameter actually does match the storage value.

Reply:
Yeah someone can do that but we can't in any way for example force all transactions to send some pre-fetched storage items to help with weight. A user can craft a custom function in decl_module to work in such a way and, well, good luck with them. I will note this maybe as an example. you can also add it yourself if you feel like it,